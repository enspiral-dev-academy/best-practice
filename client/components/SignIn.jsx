import React, { useState } from "react";

import { wrappedWith, userContext } from "../wrappers";

const formStyle = {
  display: "flex",
  flexDirection: "column",
  maxWidth: "250px",
};

export function SignIn({ signInUser }) {
  const [userData, setUserData] = useState({
    username: "",
    password: "",
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setUserData({ ...userData, [name]: value });
  };

  const handleClick = () => {
    signInUser(userData);
  };

  const { username, password } = userData;
  return (
    <>
      <h2>Sign in</h2>
      <div style={formStyle}>
        <label htmlFor="username">Username</label>
        <input
          name="username"
          id="username"
          value={username}
          onChange={handleChange}
        />

        <label htmlFor="password">Password</label>
        <input
          name="password"
          id="password"
          type="password"
          value={password}
          onChange={handleChange}
        />

        <button type="button" onClick={handleClick}>
          Sign in
        </button>
      </div>
    </>
  );
}

export default wrappedWith(userContext)(SignIn);
