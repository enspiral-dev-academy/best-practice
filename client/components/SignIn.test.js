import React from "react";
import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";

import { SignIn } from "./SignIn";

test("renders register form correctly", () => {
  const { asFragment } = render(<SignIn />);
  expect(asFragment()).toMatchSnapshot();
});

test("form field values update correctly on input", async () => {
  render(<SignIn />);

  const usernameInput = await screen.getByRole("textbox", {
    name: "Username",
  });
  const passwordInput = await screen.getByLabelText("Password");
  fireEvent.change(usernameInput, { target: { value: "test username" } });
  fireEvent.change(passwordInput, { target: { value: "test password" } });

  expect(usernameInput.value).toBe("test username");
  expect(passwordInput.value).toBe("test password");
});

test("calls signInUser on sign in button click", async () => {
  const signInUser = jest.fn();
  render(<SignIn signInUser={signInUser} />);

  const signInButton = await screen.getByRole("button", { name: "Sign in" });
  fireEvent.click(signInButton);

  expect(signInUser).toHaveBeenCalled();
});
