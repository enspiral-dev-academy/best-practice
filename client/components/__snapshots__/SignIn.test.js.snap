// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders register form correctly 1`] = `
<DocumentFragment>
  <h2>
    Sign in
  </h2>
  <div
    style="display: flex; flex-direction: column; max-width: 250px;"
  >
    <label
      for="username"
    >
      Username
    </label>
    <input
      id="username"
      name="username"
      value=""
    />
    <label
      for="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      type="password"
      value=""
    />
    <button
      type="button"
    >
      Sign in
    </button>
  </div>
</DocumentFragment>
`;
